src/backend/distributed/commands/index.c

```diff
@@ -738,6 +771,16 @@ PreprocessDropIndexStmt(Node *node, const char *dropIndexCommand,
 			continue;
 		}
 
+		/*
+		 * Acquire global lock to prevent concurrent writes or DDL. However,
+		 * we do not bother for DROP INDEX CONCURRENTLY, since we'll have
+		 * to release the lock.
+		 */
+		if (!dropIndexStatement->concurrent)
+		{
```

Comment:
hmmm, calling `AcquireDistributedLockOnRelations` after calling `RangeVarGetRelidExtended` created additional risk of deadlocks. 

Similar to the other comment, the order of acquiring the distributed locks is important. `AcquireDistributedLockOnRelations` is designed such that it is called before any lock locks are acquired. If we let local locks to be acquired before starting to acquire remote locks, then two concurrent such operations can cause deadlocks.

I'm thinking of a way that we could generalize acquiring distributed locks in a more centralized place, where we do not add the call to `AcquireDistributedLockOnRelations` but mostly add an item to a list, such as `T_LOCK_TABLE, T_CREATE_INDEX` etc.

That could potentially make the code simpler, and less prone to edge cases like this

Reply:
Currently we need to take some locks on the local node to verify that the index is actually valid. I can not think of an easy way to verify that the index exists and it is created on a distributed table without taking any locks on the local table.

We can possibly take some locks locally, and perform necessary checks, and later release all those locks before acquiring the distributed locks on relations. Does this sound good?